### PR TITLE
Developer guide for internal devs, and helper scripts

### PR DESCRIPTION
**Pull request by graebm**
* `docs/dev/README.md`: Getting started guide for contributors to aws-crt-python. I'm going to flesh this out more, or add other docs. But this is a solid start.

While writing the guide I realized helper scripts would be useful so:
* `scripts/`: new home for helper scripts. I chose to use python instead of bash so that they could be run on Windows machines as well.
* `scripts/install-dev.py`: build and install in "develop" mode
* `scripts/clean.py`: clean up misc tmp files
* `scripts/format-c.py`: format all C code
* `scripts/format-python.py`: format all Python code
* `scripts/format-all.py`: format all code
* `scripts/utils.py`: common helper functions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/dev/README.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    *   AWS_TEST_IOT_MQTT_ENDPOINT - AWS account-specific endpoint to connect to IoT core by
+    *   AWS_TEST_TLS_CERT_PATH - file path to certificate used to initialize the TLS context of the MQTT connection
+    *   AWS_TEST_TLS_KEY_PATH - file path to the private key used to initialize the TLS context of the MQTT connection
+    *   AWS_TEST_TLS_ROOT_CERT_PATH - file path to the root CA used to initialize the TLS context of the MQTT connection
+*   PKCS11
+    *   AWS_TEST_PKCS11_LIB - path to PKCS#11 library
+    *   AWS_TEST_PKCS11_PIN - user PIN for logging into PKCS#11 token
+    *   AWS_TEST_PKCS11_TOKEN_LABEL - label of PKCS#11 token
+    *   AWS_TEST_PKCS11_KEY_LABEL - label of private key on PKCS#11 token,
+        which must correspond to the cert at AWS_TEST_TLS_CERT_PATH.
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:121]: Super minor nitpick: I'd remove the newline here and do it all on one line for consistency with the rest of the file.

File: docs/dev/README.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+*   PROXY
+    *   AWS_TEST_HTTP_PROXY_HOST - host address of the proxy to use for tests that make open connections to the proxy
+    *   AWS_TEST_HTTP_PROXY_PORT - port to use for tests that make open connections to the proxy
+    *   AWS_TEST_HTTPS_PROXY_HOST - host address of the proxy to use for tests that make TLS-protected connections to the proxy
+    *   AWS_TEST_HTTPS_PROXY_PORT - port to use for tests that make TLS-protected connections to the proxy
+    *   AWS_TEST_HTTP_PROXY_BASIC_HOST - host address of the proxy to use for tests that make open connections to the proxy with basic authentication
+    *   AWS_TEST_HTTP_PROXY_BASIC_PORT - port to use for tests that make open connections to the proxy with basic authentication
+    *   AWS_TEST_BASIC_AUTH_USERNAME - username to use when using basic authentication to the proxy
+    *   AWS_TEST_BASIC_AUTH_PASSWORD - password to use when using basic authentication to the proxy
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:130]: Super minor nitpick: I'd reword this a bit to something like `username to use when connecting with basic authentication to the proxy` or `username for basic authentication use with the proxy` to avoid `to use when using`.

File: docs/dev/README.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    *   AWS_TEST_TLS_KEY_PATH - file path to the private key used to initialize the TLS context of the MQTT connection
+    *   AWS_TEST_TLS_ROOT_CERT_PATH - file path to the root CA used to initialize the TLS context of the MQTT connection
+*   PKCS11
+    *   AWS_TEST_PKCS11_LIB - path to PKCS#11 library
+    *   AWS_TEST_PKCS11_PIN - user PIN for logging into PKCS#11 token
+    *   AWS_TEST_PKCS11_TOKEN_LABEL - label of PKCS#11 token
+    *   AWS_TEST_PKCS11_KEY_LABEL - label of private key on PKCS#11 token,
+        which must correspond to the cert at AWS_TEST_TLS_CERT_PATH.
+*   PROXY
+    *   AWS_TEST_HTTP_PROXY_HOST - host address of the proxy to use for tests that make open connections to the proxy
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:123]: Super minor nitpick: I'd change this to `host address of proxy for tests that make open connections to the proxy` or even `host address of proxy for tests that make open connections`

Comment (TwistedTwigleg) [reply docs/dev/README.md:123]: Likewise, for the other properties:
    *   `AWS_TEST_HTTP_PROXY_PORT` - port for tests that make open connections to the proxy
    *   `AWS_TEST_HTTPS_PROXY_HOST` - host address for tests that make TLS-protected connections to the proxy
    *   `AWS_TEST_HTTPS_PROXY_PORT` - port for tests that make TLS-protected connections to the proxy
    *   `AWS_TEST_HTTP_PROXY_BASIC_HOST` - host address for tests that make open connections to the proxy with basic authentication
    *   `AWS_TEST_HTTP_PROXY_BASIC_PORT` - port for tests that make open connections to the proxy with basic authentication

File: docs/dev/README.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    *   AWS_TEST_HTTPS_PROXY_PORT - port to use for tests that make TLS-protected connections to the proxy
+    *   AWS_TEST_HTTP_PROXY_BASIC_HOST - host address of the proxy to use for tests that make open connections to the proxy with basic authentication
+    *   AWS_TEST_HTTP_PROXY_BASIC_PORT - port to use for tests that make open connections to the proxy with basic authentication
+    *   AWS_TEST_BASIC_AUTH_USERNAME - username to use when using basic authentication to the proxy
+    *   AWS_TEST_BASIC_AUTH_PASSWORD - password to use when using basic authentication to the proxy
+*   S3
+    *   AWS_TEST_S3 - set to any value to enable S3 tests.
+        Unfortunately, at this time these tests can only be run by members of the
+        Common Runtime team, due to use of hardcoded paths. TODO:
+        alter tests so anyone can run them.
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:135]: Super minor nitpicks: Would break this down into a bullet point with a sub-point and bold that it can only be run by CRT team so it stands out a bit more:

* `AWS_TEST_S3` - set to any value to enable S3 tests.
  * Unfortunately, at this time the tests **can only be run by members of the Common Runtime Team**, due to the use of hardcoded paths.
  * TODO: Alter tests so anyone can run them.

File: docs/dev/README.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+## Code Formatting
+
+We use automatic code formatters in this project and pull requests will fail unless
+the code is formatted correctly.
+
+`autopep8` is used for python code. You should have installed via `requirements-dev.txt` above.
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:142]: Super minor nitpick: Add `this` after `installed` so it's `You should have installed this via ...`

File: docs/dev/README.md [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            Or use the command `cmd+shift+P -> Python: Create Terminal`.
+    *   C_Cpp > Default: Include Path - ("C_Cpp.default.includePath" in json view)
+        *   Add item - set path to Python's C headers.
+            For example: `/Library/Frameworks/Python.framework/Versions/3.10/include/python3.10`
+        *   This is optional, it helps IntelliSense when viewing C files.
+
+4)  Add helpful tasks you can run via `cmd+shift+P -> Tasks: Run Task`
+    *   Copy [this file](vscode/launch.json) to `aws-crt-python/.vscode/tasks.json` for the following tasks:
+        * install - pip install in develop mode. `cmd+shift+B` is a special shortcut for this task
+        * format-python - format all .py files
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:186]: Super minor nitpick: Put command names in code blocks:
* ` install` - pip install ...
* `format-python` - format all ...

File: docs/dev/README.md [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+#### Debugging C with VSCode
+Unfortunately, we haven't figured out how to do interactive debugging of the C code.
+Python ultimately builds and links the module together, and it seems to always strip out the debug info.
+Please update this guide if you know how. For now, `printf()` is your best option.
+
+If you suspect the bug is in the external C code (i.e. [aws-c-http](https://github.com/awslabs/aws-c-http))
+and need to do interactive debugging, your best bet is cloning that external project,
+build it in Debug, and step through its tests.
+
+# TODO
```
Comment (TwistedTwigleg) [inline docs/dev/README.md:215]: A thought: Should we encourage contributors to also test to make sure the SDK builds/compiles? Like maybe for them to quickly run the pubsub just to make sure nothing there breaks?

**Review by TwistedTwigleg — APPROVED** [review]
Looks good! I added some comments on a few things, but all of them are extremely minor nitpicks and are by no means necessary. 👍

File: docs/dev/README.md [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        *   [Debugging Python](#debugging-python-with-vscode)
+        *   [Debugging C](#debugging-c-with-vscode)
+
+## Git
+
+Clone to a development folder:
+```sh
+$ git clone git@github.com:awslabs/aws-crt-python.git
+$ cd aws-crt-python
+$ git submodule update --init
```
Comment (sbSteveK) [inline docs/dev/README.md:24]: Trivial - I don't think it's necessary to use --recursive unless we've got nested submodules but it doesn't hurt to put it there in case some changes in the future to how we modularize involves nesting of submodules. It'd be kinda like futureproofing this guide but then again, maybe we never will. Figured I'd mention it because when dengke first showed me how to pull in submodules he added --recursive and I've just been using it since =P

File: docs/dev/README.md [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+Your terminal looks something like this when the virtual environment is active:
+```
+(.venv) $
+```
+Now any time you type `python3` or `pip` or even `python`, you are using the
+one from your virtual environment.
+To stop using the virtual environment, run `deactivate`.
+
+## Install
```
Comment (sbSteveK) [inline docs/dev/README.md:54]: ooc, does this install any updates only to the virtual environment you setup and activated above? If yes, should we move the installs/updates to before venv to make sure all this stuff is installed across the system? Maybe we don't care?

File: docs/dev/README.md (hunk 8 above)
Comment (TingDaoK) [inline docs/dev/README.md:24]: alternative: `git clone git@github.com:awslabs/aws-crt-python.git --recursive`?

**Review by TingDaoK — APPROVED** [review]

File: docs/dev/README.md (hunk 7 above)
Comment (graebm) [reply docs/dev/README.md:215]: nope
that is too much hassle
if we want something like that, we should build it into the CI

Seriously, it **is a good idea** to test whether it will break key downstream dependencies. But it is too much hassle to ask first time contributors to a code base to ALSO figure out how to build another package and test it, especially when you can't test the SDK without an IoT Core account, and not everyone contributing to aws-crt-python is doing IoT

File: docs/dev/README.md (hunk 8 above)
Comment (graebm) [reply docs/dev/README.md:24]: you only need the --recursive in one repo: aws-iot-device-sdk-cpp-v2

you don't need it in any of the aws-crt-xyz libraries. I'm going to keep this simple

File: docs/dev/README.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+Set up a [virtual environment](https://docs.python.org/3/library/venv.html)
+for development. This guide suggests `aws-crt-python/.venv/` as a default location.
+Create a virtual environment like so:
+```sh
+$ python3 -m venv .venv/
+```
+
+To activate the virtual environment in your current terminal:
+```sh
+$ source .venv/bin/activate
```
[the remark this answers is not in the document]
Comment (graebm) [reply docs/dev/README.md:43]: ohhh good point
I should run through this guide on windows and note any alternate steps

File: docs/dev/README.md (hunk 3 above)
Comment (graebm) [reply docs/dev/README.md:123]: I rewrote so it's not such a soup of repeating text, like:
```
    * TLS-protected connections to the proxy
        *   `AWS_TEST_HTTPS_PROXY_HOST` - proxy host address
        *   `AWS_TEST_HTTPS_PROXY_PORT` - proxy port
    * Open connections to the proxy
        *   `AWS_TEST_HTTP_PROXY_HOST` - proxy host address
        *   `AWS_TEST_HTTP_PROXY_PORT` - port port
    * Open connections to the proxy, with basic authentication
        *   `AWS_TEST_HTTP_PROXY_BASIC_HOST` - proxy host address
        *   `AWS_TEST_HTTP_PROXY_BASIC_PORT` - proxy port
        *   `AWS_TEST_BASIC_AUTH_USERNAME` - username
        *   `AWS_TEST_BASIC_AUTH_PASSWORD` - password
```

File: docs/dev/README.md (hunk 9 above)
Comment (graebm) [reply docs/dev/README.md:54]: Great suggestion though :|

Found a stackoverflow about this weirdness:
https://unix.stackexchange.com/questions/464671/why-does-pythons-pip-reset-to-version-10-0-1-in-every-new-virtual-environment
